packages/kbn-url-state/index.ts

```diff
@@ -6,4 +6,109 @@
  * Side Public License, v 1.
  */
 
-export { useSyncToUrl } from './use_sync_to_url';
+import { useCallback, useEffect, useState } from 'react';
+import { encode, decode, RisonValue } from '@kbn/rison';
+import { stringify, parse } from 'query-string';
+
+interface StateCache<T> {
+  namespaces: Record<string, Record<string, T | undefined>>;
+  timeoutHandle: number;
+}
+
+/**
+ * Temporary cache for state stored in the URL. This will be serialized to the URL
+ * in a single batched update to avoid excessive history entries.
+ */
+const cache: StateCache<unknown> = {
+  namespaces: {},
+  timeoutHandle: 0,
+};
+
+const CUSTOM_URL_EVENT = 'url:update' as const;
+
+// This is a list of events that can trigger a render.
+const URL_CHANGE_EVENTS: string[] = ['popstate', CUSTOM_URL_EVENT];
+
+/**
+ * This hook stores state in the URL, but with a namespace to avoid collisions with other values in the URL.
+ * It also batches updates to the URL to avoid excessive history entries.
+ * With it, you can store state in the URL and have it persist across page refreshes.
+ * The state is stored in the URL as a Rison encoded object.
+ *
+ * Example: when called like this `const [value, setValue] = useUrlState<boolean>('myNamespace', 'myKey');`
+ * the state will be stored in the URL like this: `?myNamespace=(myKey:!n)`
+ *
+ * State is not cleared from the URL when the hook is unmounted and this is by design.
+ * If you want it to be cleared, you can do it manually by calling `setValue(undefined)`.
+ *
+ * @param urlNamespace actual top level query param key
+ * @param key sub key of the query param
+ */
+export const useUrlState = <T = unknown>(urlNamespace: string, key: string) => {
+  if (!cache.namespaces[urlNamespace]) {
+    cache.namespaces[urlNamespace] = {};
+  }
+
+  const [internalValue, setInternalValue] = useState<T | undefined>(undefined);
```

Comment:
Since, you are already implementing this, I would like to bring one more usecase for this to your attention.

In the Filter group controls on the Alerts Page, I use a custom hook to sync the url with the state as shown below.

https://github.com/elastic/kibana/blob/6bdc094d79e65c10045711a5ae4c9813fb2e7a07/x-pack/plugins/security_solution/public/common/components/filter_group/hooks/use_control_group_sync_to_local_storage.ts#L13-L23

One requirement in this was to `pause`/`resume` sync on demand. For example, when user is editing the controls, we do not want to sync the state but otherwise, we want to.

Something to think about when finally implementing this.

Reply:
I feel like this is out of scope for this PR. Would you be ok if we take a look at this later? Especially because we might actually drop this package entirely in favor for another already existing one...

Reply:
100%, this was a use case for synching url changes that I wanted to bring to attention. No hurry on this. 

and I agree on the point of using existing one. Thanks.